include/hdlConvertor/hdlObjects/hdlTypeDec.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+#include <hdlConvertor/hdlObjects/named.h>
+#include <hdlConvertor/hdlObjects/iHdlObj.h>
+
+
+namespace hdlConvertor {
+namespace hdlObjects {
+
+/*
+ * HDL AST node for module declaration
+ * (part with ports and params for Verilog, Entity for VHDL)
+ * */
+class HdlTypeDec: public WithNameAndDoc, public iHdlObj {
+public:
+	enum TypeClass {hdltc_error, hdltc_enum, hdltc_array, hdltc_struct, 
+		hdltc_physical, hdltc_integral, hdltc_float};
+	HdlTypeDec(const std::string& name);
+	virtual TypeClass typeClass() = 0;
+	virtual ~HdlTypeDec();
+};
+
+class HdlEnumTypeDec: public HdlTypeDec {
+public:
+	std::vector<std::string> ids;
```

Comment:
Interesting.  This must be a SystemVerilog concept.  I'll try to find information online.  Can you recommend any good references?  A section of a favorite book?

In VHDL, an "enumerated" type is a full, new type.  A subtype can "inherit" from a type or subtype, but there are (almost) no built-in relationships between different types.

Reply:
Can any (sub) type have "inheritance"?  Should this inheritance be built into the base `HdlTypeDec`, or just into the `HdlEnumTypeDec`?

Reply:
> enum SystemVerilog concept

http://www.fis.agh.edu.pl/~skoczen/hdl/ieee_std/ieee1800-2005.pdf p26, `enum_base_type`

I would put it in `HdlTypeDec`.

Reply:
I've done some work with the VHDL type system, so I'm confident that VHDL doesn't need a "base type" for enumerations.  In fact, enumerations in VHDL don't have an underling integer or bit-based representation, just the declared identifiers.  (VHDL designers often use an attribute to request a particular encoding in synthesis, e.g. "onehot" or "gray" or "binary".)

Looking at the SystemVerilog enumerations, it appears that they are more like C (surprise!) and have an implicit integer encoding (type `int`, first id is 0 then ascending).  There are a few other options for the "base type" including (packed) 2-state and 4-state logic.

From this information I'm inclined to add a `base_type` (or better name) `HdlExpr` pointer for enumerations.  Since it's specific to the `enum` class, it doesn't make sense to me to put it on the base `HdlTypeDec` class.  For VHDL, this pointer will `null`, of course.  Let me know how this sounds to you.

Reply:
SystemVerilog classes should have their own inheritance mechanism.  The capabilities and use cases are different from the base type of an enum.

Reply:
OK, lets not put `base_type` to a `HdlTypeDec` as we may soon find out that there are some non-class types which can not have `base_type` (like in Java) and we would need to check the `base_type` presence to assert it is not set.